// react-facebook/module/MessengerCheckbox.js
import React from 'react';

import Parser from './Parser';

export default function MessengerCheckbox(props) {
  const {
    className,
    origin,
    prechecked,
    allowLogin,
    userRef,
    messengerAppId,
    pageId,
    children,
    size,
    onParse,
    centerAlign,
    skin
  } = props;

  return React.createElement(
    Parser,
    { className: className, onParse: onParse },
    React.createElement(
      'div',
      {
        className: 'fb-messenger-checkbox',
        messenger_app_id: messengerAppId,
        page_id: pageId,
        size: size,
        origin: origin,
        user_ref: userRef,
        prechecked: prechecked,
        allow_login: allowLogin,
        skin: skin,
        center_align: centerAlign
      },
      children
    )
  );
}

MessengerCheckbox.defaultProps = {
  size: undefined,
  allowLogin: undefined,
  prechecked: undefined,
  userRef: undefined,
  children: undefined,
  className: undefined,
  onParse: undefined,
  origin: undefined,
  skin: undefined,
  centerAlign: undefined
};
