// gatsby/dist/commands/build.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

const _ = require(`lodash`);

const path = require(`path`);

const report = require(`gatsby-cli/lib/reporter`);

const buildHTML = require(`./build-html`);

const buildProductionBundle = require(`./build-javascript`);

const bootstrap = require(`../bootstrap`);

const apiRunnerNode = require(`../utils/api-runner-node`);

const _require = require(`../utils/get-static-dir`),
      copyStaticDirs = _require.copyStaticDirs;

const _require2 = require(`../utils/tracer`),
      initTracer = _require2.initTracer,
      stopTracer = _require2.stopTracer;

const db = require(`../db`);

const chalk = require(`chalk`);

const tracer = require(`opentracing`).globalTracer();

const signalExit = require(`signal-exit`);

const telemetry = require(`gatsby-telemetry`);

const _require3 = require(`../redux`),
      store = _require3.store,
      emitter = _require3.emitter;

const queryUtil = require(`../query`);

const pageDataUtil = require(`../utils/page-data`);

const WorkerPool = require(`../utils/worker/pool`);

function reportFailure(msg, err) {
  report.log(``);
  report.panic(msg, err);
}

const waitJobsFinished = () => new Promise((resolve, reject) => {
  const onEndJob = () => {
    if (store.getState().jobs.active.length === 0) {
      resolve();
      emitter.off(`END_JOB`, onEndJob);
    }
  };

  emitter.on(`END_JOB`, onEndJob);
  onEndJob();
});

module.exports =
/*#__PURE__*/
function () {
  var _build = (0, _asyncToGenerator2.default)(function* (program) {
    const publicDir = path.join(program.directory, `public`);
    initTracer(program.openTracingConfigFile);
    telemetry.trackCli(`BUILD_START`);
    signalExit(() => {
      telemetry.trackCli(`BUILD_END`);
    });
    const buildSpan = tracer.startSpan(`build`);
    buildSpan.setTag(`directory`, program.directory);

    const _ref = yield bootstrap(Object.assign({}, program, {
      parentSpan: buildSpan
    })),
          graphqlRunner = _ref.graphqlRunner;

    const queryIds = queryUtil.calcInitialDirtyQueryIds(store.getState());

    const _queryUtil$groupQuery = queryUtil.groupQueryIds(queryIds),
          staticQueryIds = _queryUtil$groupQuery.staticQueryIds,
          pageQueryIds = _queryUtil$groupQuery.pageQueryIds;

    let activity = report.activityTimer(`run static queries`, {
      parentSpan: buildSpan
    });
    activity.start();
    yield queryUtil.processStaticQueries(staticQueryIds, {
      activity,
      state: store.getState()
    });
    activity.end();
    yield apiRunnerNode(`onPreBuild`, {
      graphql: graphqlRunner,
      parentSpan: buildSpan
    }); // Copy files from the static directory to
    // an equivalent static directory within public.

    copyStaticDirs();
    activity = report.activityTimer(`Building production JavaScript and CSS bundles`, {
      parentSpan: buildSpan
    });
    activity.start();
    const stats = yield buildProductionBundle(program).catch(err => {
      reportFailure(`Generating JavaScript bundles failed`, err);
    });
    activity.end();
    const workerPool = WorkerPool.create();
    const webpackCompilationHash = stats.hash;

    if (webpackCompilationHash !== store.getState().webpackCompilationHash) {
      store.dispatch({
        type: `SET_WEBPACK_COMPILATION_HASH`,
        payload: webpackCompilationHash
      });
      activity = report.activityTimer(`Rewriting compilation hashes`, {
        parentSpan: buildSpan
      });
      activity.start(); // We need to update all page-data.json files with the new
      // compilation hash. As a performance optimization however, we
      // don't update the files for `pageQueryIds` (dirty queries),
      // since they'll be written after query execution.

      const cleanPagePaths = _.difference([...store.getState().pages.keys()], pageQueryIds);

      yield pageDataUtil.updateCompilationHashes({
        publicDir,
        workerPool
      }, cleanPagePaths, webpackCompilationHash);
      activity.end();
    }

    activity = report.activityTimer(`run page queries`);
    activity.start();
    yield queryUtil.processPageQueries(pageQueryIds, {
      activity
    });
    activity.end();

    require(`../redux/actions`).boundActionCreators.setProgramStatus(`BOOTSTRAP_QUERY_RUNNING_FINISHED`);

    yield waitJobsFinished();
    yield db.saveState();
    activity = report.activityTimer(`Building static HTML for pages`, {
      parentSpan: buildSpan
    });
    activity.start();

    try {
      yield buildHTML.buildPages({
        program,
        stage: `build-html`,
        pagePaths: [...store.getState().pages.keys()],
        activity,
        workerPool
      });
    } catch (err) {
      reportFailure(report.stripIndent`
        Building static HTML failed${err.context && err.context.path ? ` for path "${chalk.bold(err.context.path)}"` : ``}

        See our docs page on debugging HTML builds for help https://gatsby.dev/debug-html
      `, err);
    }

    activity.end();
    yield apiRunnerNode(`onPostBuild`, {
      graphql: graphqlRunner,
      parentSpan: buildSpan
    });
    report.info(`Done building in ${process.uptime()} sec`);
    buildSpan.finish();
    yield stopTracer();
    workerPool.end();
  });

  return function build(_x) {
    return _build.apply(this, arguments);
  };
}();
